docs/rfcs/20220209_leader_epoch_support_rfc.md

```diff
@@ -0,0 +1,114 @@
+- Feature Name: Leader epoch handling in redpanda
+- Status: in-progress
+- Start Date: 2022-02-09
+- Authors: Michal Maslanka
+- Issue: https://github.com/redpanda-data/redpanda/issues/3689
+
+# Executive Summary
+
+In kafka some of the client Requests are fenced with the Leader Epoch. 
+Leader Epoch is a 32-bit integer that is increased every time partition 
+leadership changes. The Leader Epoch is used to detect stale metadata, 
+fence produce and fetch request and recognize truncations.
+
+## What is being proposed
+
+Using Raft protocol `term` as a Kafka leader epoch. The `term` is semantically 
+identical to Kafka's leader epoch as it is increased every time Raft group 
+leadership changes and stays constant when leader is active.
+
+## Why (short reason)
+
+Leader epoch is used by Kafka clients to detect truncations and stale metadata. 
+In Redpanda we were missing that mechanism. This may lead to some issues that 
+may occur when one of the nodes is isolated and do not receive metadata updates. 
+Returning leader epoch will allow the client to identify this situation 
+and update metadata against different broker. 
+
+## How (short plan)
+
+We will expose raft group terms, that are already gathered in redpanda metadata 
+infrastructure as leader epochs. Additionally we will implement 
+`OffsetForLeaderEpoch` Kafka API to provide full compatibility. 
+
+
+# Motivation
+## Why are we doing this?
+
+Adding leader epoch support will bring allow us to better handle stale metadata 
+and unclean truncations.
+
+### Note on unclean truncations in redpanda
+
+Since Redpanda replication is based on Raft we do not allow for unclean leader 
+election that may lead to high watermark being decreased. The only situation in 
+which high watermark may be decreased is simultaneous failure of majority of 
+nodes handling relaxed consistency writes such as `ACKS=1` and `ACKS=0`. 
```

Comment:
```suggestion
Since Redpanda replication is based on Raft, we do not allow for unclean leader 
election that may lead to the high watermark being decreased. The only situation in 
which the high watermark may be decreased is simultaneous failure of the majority of 
nodes handling relaxed consistency writes such as `ACKS=1` and `ACKS=0`. 
```